cmd/maya-apiserver/app/server/volume_endpoint_v1alpha1.go

```diff
@@ -57,56 +55,196 @@ func (v *volumeAPIOpsV1alpha1) get() (interface{}, error) {
+	hdrNS := ""
+
+	// get volume related details from http request
+	if v.req != nil {
+		decodeBody(v.req, vol)
+		hdrNS = v.req.Header.Get(NamespaceKey)
+	}
+
+	vol.Name = volumeName
+
+	// volume name is expected
+	if len(vol.Name) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("failed to read volume: missing volume name '%v'", vol))
+	}
+
+	// use namespace from labels if volume ns is not set
+	if len(vol.Namespace) == 0 {
+		vol.Namespace = vol.Labels[string(v1alpha1.NamespaceCVK)]
+	}
+
+	// use namespace from req headers if volume ns is still not set
+	if len(vol.Namespace) == 0 {
+		vol.Namespace = hdrNS
+	}
+
+	vOps, err := volume.NewVolumeOperation(vol)
+	if err != nil {
+		return nil, CodedError(400, err.Error())
+	}
+
+	cvol, err := vOps.Read()
+	if err != nil {
+		glog.Errorf("failed to read cas template based volume: error '%s'", err.Error())
+		return nil, CodedError(500, err.Error())
+	}
+
+	glog.Infof("cas template based volume was read successfully: name '%s'", cvol.Name)
+	return cvol, nil
 }
 
 func (v *volumeAPIOpsV1alpha1) delete(volumeName string) (interface{}, error) {
-	return nil, fmt.Errorf("Not implemented")
+	glog.Infof("cas template based volume delete request was received")
+
+	vol := &v1alpha1.CASVolume{}
+	hdrNS := ""
+
+	// get volume related details from http request
+	if v.req != nil {
+		decodeBody(v.req, vol)
+		hdrNS = v.req.Header.Get(NamespaceKey)
+	}
+
+	vol.Name = volumeName
+
+	// volume name is expected
+	if len(vol.Name) == 0 {
+		return nil, CodedError(400, fmt.Sprintf("failed to delete volume: missing volume name '%v'", vol))
+	}
+
+	// use run namespace from labels if volume ns is not set
+	if len(vol.Namespace) == 0 {
+		vol.Namespace = vol.Labels[string(v1alpha1.NamespaceCVK)]
+	}
+
+	// use run namespace from req headers if volume ns is still not set
+	if len(vol.Namespace) == 0 {
+		vol.Namespace = hdrNS
+	}
+
+	vOps, err := volume.NewVolumeOperation(vol)
+	if err != nil {
+		return nil, CodedError(400, err.Error())
+	}
+
+	cvol, err := vOps.Delete()
+	if err != nil {
+		glog.Errorf("failed to delete cas template based volume: error '%s'", err.Error())
+		return nil, CodedError(500, err.Error())
+	}
+
+	glog.Infof("cas template based volume was deleted successfully: name '%s'", cvol.Name)
+	return cvol, nil
 }
 
 func (v *volumeAPIOpsV1alpha1) list() (interface{}, error) {
-	return nil, fmt.Errorf("Not implemented")
+	glog.Infof("cas template based volume list request was received")
+
+	vols := &v1alpha1.CASVolumeList{}
+	hdrNS := ""
+
+	// extract volume list details from http request
+	if v.req != nil {
+		decodeBody(v.req, vols)
+		hdrNS = v.req.Header.Get(NamespaceKey)
+	}
+
+	// use run namespace from labels if volume ns is not set
+	if len(vols.Namespace) == 0 {
+		vols.Namespace = vols.Labels[string(v1alpha1.NamespaceCVK)]
+	}
+
+	// use run namespace from req headers if volume ns is still not set
+	if len(vols.Namespace) == 0 {
+		vols.Namespace = hdrNS
+	}
```

Comment:
This stuff is getting used in each CRUD API request , it will be good to have a method to get the Namespace. ?

Reply:
Will refactor in future PR... There will be more changes here due to upcoming changes in provisioner.